app/models/katello/glue/provider.rb

```diff
@@ -144,14 +150,39 @@ def import_product(product_json)
         if product&.redhat?
           product.update!(:name => product_json['name']) unless product.name == product_json['name']
         elsif product.nil?
-          product = Glue::Candlepin::Product.import_from_cp(product_json, organization)
+          # we have an orphaned Red Hat product
+          product = Glue::Candlepin::Product.import_redhat_product_from_cp(product_json, organization)
         end
         product
       end
 
-      def index_subscriptions
-        Katello::Subscription.import_all(organization)
-        Katello::Pool.import_all(organization, false)
+      def orphaned_custom_product?(product_json, organization)
+        return unless orphaned_product?(product_json, organization)
+        product_provider = provider_for_cp_id(cp_id: product_json['id'], organization: organization)
+        unless product_provider.redhat_provider?
+          Rails.logger.info "Found orphaned custom product #{product_json['id']} in Candlepin: '#{product_json['name']}'. Skipping import into Katello; run rake katello:delete_orphaned_custom_products to remove it from Candlepin."
+          true
+        else
+          false
+        end
+      end
+
+      def orphaned_product?(product_json, organization)
+        organization.products.find_by(:cp_id => product_json['id']).blank?
```

Comment:
If you need to query this per-product, consider using a `where(...).exists?` this way you avoid unnecessarily instantiating the product if it does exist